components/script/dom/cssstyledeclaration.rs

```diff
@@ -218,14 +246,17 @@ impl CSSStyleDeclarationMethods for CSSStyleDeclaration {
                         declarations: declarations.into_iter().map(|d| (d, importance)).collect(),
                         important_count: important_count,
                     };
-                    self.owner.set_style_attr(block.to_css_string());
-                    *option = Some(Arc::new(RwLock::new(block)));
+                    if let CSSStyleOwner::Element(ref el) = self.owner {
+                        el.set_style_attr(block.to_css_string());
+                        *el.style_attribute().borrow_mut() = Some(Arc::new(RwLock::new(block)));
+                    }
                 }
             }
         }
 
-        let node = self.owner.upcast::<Node>();
-        node.dirty(NodeDamage::NodeStyleDamaged);
+        if let CSSStyleOwner::Element(ref el) = self.owner {
+            el.upcast::<Node>().dirty(NodeDamage::NodeStyleDamaged);
```

Comment:
For `Element.style`, that element. For anything in a stylesheet, conservatively, the entire document. (Maybe this can be refined, e.g. changing rules inside a not-currently-matching `@media` rule doesn’t need to mark anything dirty.)